### ISSUE TITLE
Slow training...

### ISSUE THREAD

**Issue by maff20** [freewym/espresso#41]
Hello,

I have spent some time to compare pychain LF-MMI in Espresso and the pychain_example, which seems to borrow some code from Espresso. I get very slow forward passes in Espresso while they are much faster in PyChain_example (I use DistributedDataParallel on both Espresso ('no_c10d' backend, which uses NCCL anyway?) and PyChain (with 'nccl') ). I use a TDNN model which is the same, same architecture cnn/bn/relu implementation matched from Espresso to PyChain: 6 TDNN+BN+ReLU layers, strides=(1,1,1,1,1,3), dilation=(1,1,1,3,3,3), kernels=(3,3,3,3,3,3), no residual connections. Both use curriliculum learning in the first epoch and start with shortest batches. 

Espresso code:
```
        start= time.time()
        for i in range(len(self.tdnn)):
            if self.residual and i > 0:  # residual connection starts from the 2nd layer
                prev_x = x
            x, x_lengths, padding_mask = self.tdnn[i](x, x_lengths)
            x = self.dropout_out_module(x)
            x = x + prev_x if self.residual and i > 0 and x.size(1) == prev_x.size(1) else x
        print ('6xTDNN time %.5fs' % (time.time() - start,), 'tensor_in_size', s, 'gpu', x.get_device())
```
PyChain code:
```
        start = time.time()
        for i in range(len(self.tdnn)):
            if self.residual and i>0:
              x_prev = x
            x, x_lengths = self.tdnn[i](x, x_lengths)
            x = F.dropout(x, p=self.dropout, training=self.training)
            if self.residual and i>0 and x.size(1)==x_prev.size(1):
                x += x_prev
        print ('6xTDNN time %.5fs' % (time.time() - start,), 'tensor_in_size', s, 'gpu', x.get_device())
```
So, the code is almost line-by-line the same, architecture is the same. Yet, after using DistributedDataParallel, Espresso is much  slower. This was run on the same machine, same 2 GPUs, one experiment right after the other (so no load change issues on the machine). I checked that the computing the padding does not significantly affect the timing. Here are the timings for several forward passess of similar size.

Espresso:
6xTDNN time 2.42642s tensor_in_size torch.Size([64, 158, 40]) tensor_out_size torch.Size([64, 53, 640]) gpu 1
6xTDNN time 2.39317s tensor_in_size torch.Size([64, 177, 40]) tensor_out_size torch.Size([64, 59, 640]) gpu 1
6xTDNN time 1.95155s tensor_in_size torch.Size([64, 144, 40]) tensor_out_size torch.Size([64, 48, 640]) gpu 0
6xTDNN time 2.50637s tensor_in_size torch.Size([64, 170, 40]) tensor_out_size torch.Size([64, 57, 640]) gpu 0
6xTDNN time 1.79735s tensor_in_size torch.Size([64, 192, 40]) tensor_out_size torch.Size([64, 64, 640]) gpu 1
6xTDNN time 2.37481s tensor_in_size torch.Size([64, 186, 40]) tensor_out_size torch.Size([64, 62, 640]) gpu 0

...
PyChain:
6xTDNN time 0.07956s tensor_in_size torch.Size([64, 170, 40]) tensor_out_size torch.Size([64, 57, 640]) gpu 0
6xTDNN time 0.08923s tensor_in_size torch.Size([64, 194, 40]) tensor_out_size torch.Size([64, 65, 640]) gpu 1
6xTDNN time 0.08312s tensor_in_size torch.Size([64, 211, 40]) tensor_out_size torch.Size([64, 71, 640]) gpu 0
6xTDNN time 0.08275s tensor_in_size torch.Size([64, 224, 40]) tensor_out_size torch.Size([64, 75, 640]) gpu 1
6xTDNN time 0.08598s tensor_in_size torch.Size([64, 233, 40]) tensor_out_size torch.Size([64, 78, 640]) gpu 0
6xTDNN time 0.08788s tensor_in_size torch.Size([64, 241, 40]) tensor_out_size torch.Size([64, 81, 640]) gpu 1
...
So, PyChain is 10-20 times faster... Espresso uses 40-50% of each GPU, while PyChain uses 85-95% when put together with the LF-MMI loss. I wonder how to make Espresso train as fast as PyChain shows it is possible. Is it a matter of the DistributedDataParallel imlementation in Fairseq? the backend? Any help is welcome.

**Comment by maff20** [comment]
I think I found the issue. Espresso was using a different PyTorch distribution and the Conv1d dilated convolutions were extremely slow (probably running on CPU?). I'll check this out. Sorry for raising an unnecessary issue here.

**Comment by freewym** [comment]
How did Espresso and PyChain_example compare when using 1 GPU? Are they close? I remember their speeds are close on my side.

What do you mean by "a different PyTorch distribution"?

**Comment by maff20** [comment]
Here are the timings for 1 GPU:

Espresso:
6xTDNN time 1.79834s tensor_in_size torch.Size([64, 158, 40]) tensor_out_size torch.Size([64, 53, 640]) gpu 0
6xTDNN time 1.73702s tensor_in_size torch.Size([64, 170, 40]) tensor_out_size torch.Size([64, 57, 640]) gpu 0
6xTDNN time 1.74261s tensor_in_size torch.Size([64, 177, 40]) tensor_out_size torch.Size([64, 59, 640]) gpu 0
6xTDNN time 1.79291s tensor_in_size torch.Size([64, 186, 40]) tensor_out_size torch.Size([64, 62, 640]) gpu 0
6xTDNN time 1.72016s tensor_in_size torch.Size([64, 192, 40]) tensor_out_size torch.Size([64, 64, 640]) gpu 0
6xTDNN time 1.77084s tensor_in_size torch.Size([64, 198, 40]) tensor_out_size torch.Size([64, 66, 640]) gpu 0

PyChain:
6xTDNN time 0.08386s tensor_in_size torch.Size([64, 170, 40]) tensor_out_size torch.Size([64, 57, 640]) gpu 0
6xTDNN time 0.07439s tensor_in_size torch.Size([64, 194, 40]) tensor_out_size torch.Size([64, 65, 640]) gpu 0
6xTDNN time 0.09271s tensor_in_size torch.Size([64, 211, 40]) tensor_out_size torch.Size([64, 71, 640]) gpu 0
6xTDNN time 0.08501s tensor_in_size torch.Size([64, 224, 40]) tensor_out_size torch.Size([64, 75, 640]) gpu 0
6xTDNN time 0.09431s tensor_in_size torch.Size([64, 233, 40]) tensor_out_size torch.Size([64, 78, 640]) gpu 0
6xTDNN time 0.09703s tensor_in_size torch.Size([64, 241, 40]) tensor_out_size torch.Size([64, 81, 640]) gpu 0

Timings are comparable to mutli-GPU case.

I timed the different layers in the fwd pass for Espresso. It seems that TDNN using dilated convolution (3, 4 and 5) are the issue, taking super long time compared to 0,1 and 2, which do not use dilation.

layer 1: dilation=1
transpose 1,2 time 0.00007s gpu 0
tdnn time 0.00071s gpu 0
bn time 0.00273s gpu 0
relu time 0.00222s gpu 0
transpose 2,1 time 0.00238s gpu 0
transpose 1,2 time 0.00065s gpu 0
1xTDNN layer 1 time 0.00817s gpu 0

layer 4: dilation=3
transpose 1,2 time 0.00003s gpu 1
tdnn time 0.55005s gpu 0
bn time 0.00036s gpu 0
relu time 0.00022s gpu 0
transpose 1,2 time 0.00023s gpu 0
transpose 2,1 time 0.00023s gpu 0
1xTDNN layer 4 time 0.55152s gpu 0

So, it seems clear now the slow speed comes from Conv1d with dilation alone, on 1 or 2 GPU cases.

I have different Python environments for PyChain and Espresso. Probably not well setup in the case of Espresso. Dilation should not affect timings that dramatically, if at all, I would think. Is there a way I can check CuDNN is being used?

**Comment by maff20** [comment]
After setting:

```
torch.backends.cudnn.deterministic = False
```

in speech_recognition_hybrid.py, timings are fast now. Maybe it is related to how PyTorch chooses the best CuDNN functions or CUDA kernels?

Timings now for Espresso:

6xTDNN time 0.02708s tensor_in_size torch.Size([64, 177, 40]) tensor_out_size torch.Size([64, 59, 640]) gpu 1
6xTDNN time 0.03717s tensor_in_size torch.Size([64, 170, 40]) tensor_out_size torch.Size([64, 57, 640]) gpu 0
6xTDNN time 0.02872s tensor_in_size torch.Size([64, 192, 40]) tensor_out_size torch.Size([64, 64, 640]) gpu 1
6xTDNN time 0.03489s tensor_in_size torch.Size([64, 186, 40]) tensor_out_size torch.Size([64, 62, 640]) gpu 0
6xTDNN time 0.03025s tensor_in_size torch.Size([64, 203, 40]) tensor_out_size torch.Size([64, 68, 640]) gpu 1
6xTDNN time 0.03281s tensor_in_size torch.Size([64, 198, 40]) tensor_out_size torch.Size([64, 66, 640]) gpu 0

which is even faster than PyChain! GPU utilization on 2 GPUs is now consistently around 90-100%!

**Comment by maff20** [comment]
Timings for one batch's 6xTDNN forward pass

cudnn.deterministic=True
cudnn.benchmark=True
```
2.05270s
1.94792s
1.91263s
2.54994s
```

cudnn.deterministic=False
cudnn.benchmark=True
````
0.07221s
0.07651s
0.08093s
0.09113s
```

cudnn.deterministic = False
cudnn.benchmark = False
```
0.03088s
0.03711s
0.03569s
0.03868s
```

Somehow CuDNN's autotune can't guess the best algorithms for dilated convolution? I guess the variable batch size must be a factor here too.

**Comment by freewym** [comment]
OK. Good catch! It looks like the speed gain is mainly from setting cudnn.deterministic to False, and I think cudnn.benchmark=False by default. It is said that when cudnn.benchmark=True, PyTorch will try to predict the best alogorithm every time when the input size gets changed, adding more overhead time.

Regarding the cudnn.deterministic thing, I am not sure if it is fixable. Maybe request PyTorch team to fix that.
